src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
 #ifdef FEATURE_SIMD
-        // Return a dummy node, as the return is already removed.
-        if (varTypeIsSIMD(nodeTy))
-        {
-            nodeTy = TYP_DOUBLE;
-        }
+            // Return a dummy node, as the return is already removed.
+            if (varTypeIsSIMD(nodeTy))
+            {
+                nodeTy = TYP_DOUBLE;
+            }
 #endif
-        result = gtNewZeroConNode(genActualType(nodeTy));
-        result = fgMorphTree(result);
+            result = gtNewZeroConNode(genActualType(nodeTy));
+            result = fgMorphTree(result);
+        }
     }
 
     return result;
 }
 
 /*****************************************************************************
  *
- *  Transform the given GT_CALL tree for tail call code generation.
+ * Transform the given GT_CALL tree for tail call code generation.
+ * This transforms
+ *   CALL
+ *       {callTarget}
+ *       {this}
+ *       {args}
+ * to
+ *   COMMA
+ *     CALL StoreArgsStub
+ *       {callTarget} // depending on flags provided by the runtime
+ *       {this}       // as a regular arg
+ *       {args}
+ *     COMMA
+ *       CALL Dispatcher
+ *         ADDR ReturnAddress
+ *         CallTargetStub
+ *         ADDR ReturnValue
+ *       LCL ReturnValue
+ * whenever the call node returns a value. If the call node does not return a
+ * value the last comma will not be there.
  */
-void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, void* pfnCopyArgs)
+GenTree* Compiler::fgMorphTailCallViaHelpers(GenTreeCall* call, CORINFO_TAILCALL_HELPERS& help)
 {
-#if defined(_TARGET_UNIX_)
-    noway_assert(!"Slow tail calls not supported on non-Windows platforms.");
-#endif
-
-    JITDUMP("fgMorphTailCallViaHelper (before):\n");
+    JITDUMP("fgMorphTailCallViaHelpers (before):\n");
     DISPTREE(call);
 
-    // The runtime requires that we perform a null check on the `this` argument before
-    // tail calling  to a virtual dispatch stub. This requirement is a consequence of limitations
-    // in the runtime's ability to map an AV to a NullReferenceException if
-    // the AV occurs in a dispatch stub that has unmanaged caller.
+    // TODO: Some comments (check old document and see what still applies)
+
+    // Don't support tail calling helper methods
+    assert(call->gtCallType != CT_HELPER);
+
+    // We come this route only for tail prefixed calls that cannot be dispatched as
+    // fast tail calls
+    assert(!call->IsImplicitTailCall());
+    assert(!fgCanFastTailCall(call, nullptr));
+
+    // If VSD then get rid of arg to VSD since we turn this into a direct call.
+    // The extra arg will be the first arg so this needs to be done before we
+    // handle the retbuf below.
     if (call->IsVirtualStub())
     {
-        call->gtFlags |= GTF_CALL_NULLCHECK;
+        JITDUMP("This is a VSD\n");
+// X86/ARM32 do not include the stub arg in the arg list.
+#if !defined(_TARGET_X86_) && !defined(_TARGET_ARM_)
+        call->gtCallArgs = call->gtCallArgs->GetNext();
```

Comment:
These seem pretty dangerous as the order matters. If we ever change them it is possible this could create bugs that may go unnoticed in the change they are made. Can you add asserts that the args you are removing are actually the args you expect to remove.

Reply:
Do you have a suggestion on how to check this?

Reply:
I think that in some cases we should at least be able to check the types